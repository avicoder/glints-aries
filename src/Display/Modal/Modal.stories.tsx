import React, { useState } from 'react';
import { Story, Meta } from '@storybook/react';
import { times, constant } from 'lodash';

import { Modal, Props } from './Modal';
import { Button } from '../../General/Button/Button';
import { BaseContainer } from '../../Layout/GlintsContainer/GlintsContainer';

export default {
  title: 'Display/Modal',
  component: Modal,
  decorators: [Story => <BaseContainer>{Story()}</BaseContainer>],
} as Meta;

const defaultTitle = 'Lorem Ipsum';
const defaultChildren =
  'Lorem ipsum dolor sit amet consectetur adipisicing elit.';

interface ButtonStoryProps extends Props {
  buttonTexts?: string[];
}

const Template: Story<ButtonStoryProps> = ({
  children,
  title,
  buttonTexts = [],
  ...args
}) => {
  const [isModalVisible, setIsModalVisible] = useState(false);
  return (
    <>
      <Button
        variant="solid-blue"
        onClick={() => setIsModalVisible(!isModalVisible)}
      >
        Toggle
      </Button>
      <Modal
        {...args}
        title={title || defaultTitle}
        isVisible={isModalVisible}
        onClose={() => setIsModalVisible(false)}
        footer={[
          <Button
            variant="ghost"
            onClick={() => setIsModalVisible(false)}
            key="cancel"
          >
            {buttonTexts[0] || 'Cancel'}
          </Button>,
          <Button
            variant="solid-blue"
            onClick={() => setIsModalVisible(false)}
            key="save"
          >
            {buttonTexts[1] || 'Save'}
          </Button>,
        ]}
      >
        {children || defaultChildren}
      </Modal>
    </>
  );
};

export const Interactive = Template.bind({});
Interactive.args = {
  title: defaultTitle,
  children: defaultChildren,
};

export const ScreenCentered = Template.bind({});
ScreenCentered.args = {
  centering: true,
};

export const ScreenCenteredWithALotOfContent = Template.bind({});
ScreenCenteredWithALotOfContent.args = {
  centering: true,
  children: times(10, constant(defaultChildren)).join(' '),
};

export const FooterResponsiveness = Template.bind({});
FooterResponsiveness.args = {
  buttonTexts: ['Looooooong Cancel', 'Loooooooong Save'],
  size: 's',
};
FooterResponsiveness.parameters = {
  docs: {
    description: {
      story:
        'If all the buttons cannot fit in one row, then they will be expanded to the 100% width of the container and break into two rows.',
    },
  },
};
