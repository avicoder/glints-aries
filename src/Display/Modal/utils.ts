import * as React from 'react';

/*
 * This hook is used to detect if element's children are displayed in multi-lines.
 * Currently, we use it to detect ModalFooter's buttons with "flex-wrap: wrap"
 * to responsively change style.
 */
export const checkIsChildrenInMultiLines = (
  ref: React.RefObject<HTMLDivElement>
) => {
  if (!ref.current) {
    return false;
  }

  const childNodes = ref.current.childNodes;

  if (childNodes.length < 2) {
    return false;
  }

  const firstElement = childNodes[0] as HTMLElement;
  const secondElement = childNodes[1] as HTMLElement;

  return firstElement.offsetTop !== secondElement.offsetTop;
};
