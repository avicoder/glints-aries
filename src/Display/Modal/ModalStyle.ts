import styled from 'styled-components';
import { Device } from '../../Utils/StyleConfig';
import { Greyscale } from '../../Utils/Colors';
import { Props } from './Modal';

export const SIZES: { [s: string]: number } = {
  s: 300,
  m: 500,
  l: 800,
  xl: 920,
  default: 500,
};

interface InnerProps extends Omit<Props, 'isVisible'> {
  isOpen: boolean;
}

export const ModalContainer = styled.div<InnerProps>`
  visibility: ${({ isOpen }) => (isOpen ? 'visible' : 'hidden')};
  position: fixed;
  display: flex;
  justify-content: center;
  align-items: ${({ centering }) => centering && 'center'};
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  z-index: 1001;
  background: rgba(0, 0, 0, 0.65);
  overflow: auto;

  ${({ removeAnimation, isOpen }) => {
    if (!removeAnimation) {
      return `
        opacity: ${isOpen ? '1' : '0'};
        transform: ${isOpen ? 'scale(1)' : 'scale(0.98)'};
        transform-origin: center;
        transition: ${isOpen ? 'all .2s ease-in' : 'all .15s ease-out'};
      `;
    }
  }}
`;

export const ModalDialog = styled.div`
  position: relative;
`;

export const ModalHeader = styled.header`
  display: flex;
  justify-content: space-between;
  align-items: center;
  align-content: center;
  font-size: 1.6em;
  border-bottom: 1px solid ${Greyscale.lightgrey};
  padding: 16px;

  h3 {
    font-size: 0.8em;
    text-transform: uppercase;
    margin: 0;
    margin-right: 36px;
  }

  button {
    display: flex;
    background: transparent;
    border: none;
    cursor: pointer;
    padding: 0;
    width: 15px;
  }
`;

export const ModalBody = styled.section<ModalBodyProps>`
  position: relative;
  ${({ centering }) => {
    if (centering) {
      return `
        max-height: 300px;
        overflow: auto;
      `;
    }
  }}
  padding: ${({ hideContentArea }) => (hideContentArea ? '0' : '20px 30px')};

  @media ${Device.mobileM} {
    padding: ${({ hideContentArea }) => (hideContentArea ? '0' : '20px 16px')};
  }
`;

interface ModalBodyProps {
  hideContentArea: boolean;
  centering: boolean;
}

export const ModalFooter = styled.footer<{ isChildrenInMultiLines: boolean }>`
  position: relative;
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
  padding: 16px 30px;
  justify-content: flex-end;
  border-top: 1px solid ${Greyscale.lightgrey};

  @media ${Device.mobileM} {
    padding: 16px;
  }

  ${({ isChildrenInMultiLines }) => {
    if (isChildrenInMultiLines) {
      return `
        flex-direction: column-reverse;

        > * {
          flex-grow: 1;

          &:nth-child(2) {
            margin-bottom: 10px; 
          }

          button {
            width: 100%;
          }
        }
      `;
    }
    return `
      > *:nth-child(2) {
        margin-left: 10px;
      }
    `;
  }}
`;

export const ModalContentArea = styled.div<InnerProps>`
  position: relative;
  background: ${({ hideContentArea }) =>
    hideContentArea ? 'transparent' : `${Greyscale.white}`};
  margin: 120px auto;
  max-width: 95vw;
  outline: none;

  ${({ size }) => {
    switch (size) {
      case 's':
        return `
          width: ${SIZES.s}px;
        `;
      case 'm':
        return `
          width: ${SIZES.m}px;
        `;
      case 'l':
        return `
          width: ${SIZES.l}px;
        `;
      case 'xl':
        return `
          width: ${SIZES.xl}px;
        `;
      default:
        return `
          width: ${SIZES.default}px;
        `;
    }
  }}

  ${({ centering }) => {
    if (centering) {
      return `
        display: flex;
        flex-direction: column;
        max-height: 85vh;
      `;
    }
  }}

  ${({ fullscreen }) => {
    if (fullscreen) {
      return `
        display: flex;
        flex-direction: column;
        max-width: none;
        max-height: none;
        min-width: 100vw;
        min-height: 100vh;
        margin: 0;

        ${ModalBody} {
          overflow: auto;
        }

        ${ModalFooter} {
          margin-top: auto;
        }
      `;
    }
  }}
    
  @media ${Device.mobileM} {
    width: 95vw;
  }

  ${({ removeAnimation, isOpen }) => {
    if (!removeAnimation) {
      return `
        opacity: ${isOpen ? '1' : '0'};
        transform: ${
          isOpen
            ? 'matrix3d(1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1)'
            : 'matrix3d(1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 30, 0, 1)'
        };
        transition: ${isOpen ? 'all .25s ease-in' : 'all .25s ease-out'};
      `;
    }
  }}
`;
