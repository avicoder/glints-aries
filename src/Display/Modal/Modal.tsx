import React, {
  HTMLAttributes,
  FC,
  useRef,
  useState,
  useLayoutEffect,
  useEffect,
  useCallback,
  ReactNode,
  ReactElement,
} from 'react';
import classNames from 'classnames';
import debounce from 'lodash/debounce';
import {
  disableBodyScroll,
  enableBodyScroll,
  clearAllBodyScrollLocks,
} from 'body-scroll-lock';

import { CloseIcon } from '../../General/Icon/components';
import { escEvent as createEscapeKeyEventListener } from '../../Utils/DomUtils';
import { checkIsChildrenInMultiLines } from './utils';
import {
  ModalContainer,
  ModalDialog,
  ModalContentArea,
  ModalHeader,
  ModalBody,
  ModalFooter,
} from './ModalStyle';

export const Modal: FC<Props> = ({
  isVisible,
  title,
  onClose,
  children,
  className,
  hideContentArea,
  centering,
  removeAnimation,
  footer,
  size,
  hideHeader,
  ...restProps
}) => {
  const modalContentAreaRef = useRef(null);
  const modalFooterRef = useRef(null);
  const [
    isFooterChildrenInMultiLines,
    setIsFooterChildrenInMultiLines,
  ] = useState(false);

  const handleClose = React.useCallback(() => {
    if (typeof onClose === 'function') {
      onClose();
    }
  }, [onClose]);

  useLayoutEffect(() => {
    if (!modalContentAreaRef.current) return;

    if (isVisible) {
      // On modal open
      modalContentAreaRef.current.focus();
      disableBodyScroll(modalContentAreaRef.current);
    } else {
      // On modal close
      enableBodyScroll(modalContentAreaRef.current);
    }
    return () => {
      clearAllBodyScrollLocks();
    };
  }, [isVisible, modalContentAreaRef]);

  useEffect(() => {
    const escapeKeyEventListener = createEscapeKeyEventListener(() => {
      if (isVisible) handleClose();
    });

    document.addEventListener('keydown', escapeKeyEventListener, false);

    // This function will be called on unmount and _before_ this effect is
    // re-executed because its dependencies changed. This gives us the chance
    // to clean up the existing escape key event listener.
    return () =>
      document.removeEventListener('keydown', escapeKeyEventListener, false);
  }, [isVisible, handleClose]);

  useLayoutEffect(
    function checkFooterResponsiveStyleOnMountAndOnWindowResize() {
      const checkIsFooterChildrenInMultiLines = () => {
        const isChildrenInMultiLines = checkIsChildrenInMultiLines(
          modalFooterRef
        );
        setIsFooterChildrenInMultiLines(isChildrenInMultiLines);
      };
      const debouncedCheckFooter = debounce(
        checkIsFooterChildrenInMultiLines,
        500
      );
      debouncedCheckFooter();
      window.addEventListener('resize', debouncedCheckFooter);
      return () => window.removeEventListener('resize', debouncedCheckFooter);
    },
    []
  );

  // To prevent the modal from closing
  // when a mousedown event occurs inside the ModalContentArea
  // but the subsequent mouseup event occurs outside
  const mouseDownTarget = useRef<HTMLElement>(null);

  const handleMouseDown = useCallback(
    (e: React.MouseEvent<HTMLDivElement, MouseEvent>) => {
      const element = e.target as HTMLElement;
      mouseDownTarget.current = element;
    },
    [mouseDownTarget]
  );

  const handleClick = useCallback(
    (e: React.MouseEvent<HTMLDivElement, MouseEvent>) => {
      const element = e.target as HTMLElement;
      if (mouseDownTarget.current === element) {
        handleClose();
      }
    },
    [mouseDownTarget, handleClose]
  );

  return (
    <ModalContainer
      data-testid="modal-container"
      className={classNames('aries-modal', className)}
      centering={centering}
      isOpen={isVisible}
      removeAnimation={removeAnimation}
      onMouseDown={handleMouseDown}
      onClick={handleClick}
    >
      <ModalDialog className="modal-dialog">
        <ModalContentArea
          className="modal-content"
          data-testid="dialog"
          role="dialog"
          aria-modal="true"
          hideContentArea={hideContentArea}
          centering={centering}
          onClick={e => e.stopPropagation()}
          tabIndex={0}
          isOpen={isVisible}
          removeAnimation={removeAnimation}
          size={size}
          ref={modalContentAreaRef}
          {...restProps}
        >
          {!hideHeader && (
            <ModalHeader className="modal-header">
              <h3>{title}</h3>
              <button
                aria-label="Close button"
                data-testid="close-button"
                type="button"
                onClick={handleClose}
              >
                <CloseIcon color={hideContentArea ? 'white' : 'grey'} />
              </button>
            </ModalHeader>
          )}
          <ModalBody
            className="modal-body"
            hideContentArea={hideContentArea}
            centering={centering}
          >
            {isVisible && children}
          </ModalBody>
          {footer !== undefined && (
            <ModalFooter
              ref={modalFooterRef}
              className="modal-footer"
              isChildrenInMultiLines={isFooterChildrenInMultiLines}
            >
              {footer}
            </ModalFooter>
          )}
        </ModalContentArea>
      </ModalDialog>
    </ModalContainer>
  );
};

export type sizeType = 's' | 'm' | 'l' | 'xl';

export interface Props extends Omit<HTMLAttributes<HTMLDivElement>, 'title'> {
  title?: ReactNode; // we don't really use this title attribute for div element
  isVisible: boolean;
  onClose?(): void;
  hideContentArea?: boolean;
  centering?: boolean;
  removeAnimation?: boolean;
  /** An array containing 0-2 elements. */
  footer?: ReactElement[];
  size?: sizeType;
  hideHeader?: boolean;
}

Modal.defaultProps = {
  onClose: () => undefined as () => void,
  size: 'm',
  hideContentArea: false,
  hideHeader: false,
  removeAnimation: false,
  centering: false,
};

export default Modal;
