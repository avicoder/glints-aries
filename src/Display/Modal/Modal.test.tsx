import * as React from 'react';

import * as renderer from 'react-test-renderer';
import '@testing-library/jest-dom/extend-expect';
import { fireEvent, render } from '@testing-library/react';

import Modal, { sizeType, Props } from './Modal';
import { SIZES } from './ModalStyle';

const props = {
  title: 'Modal Title',
  content: 'Modal Content',
  footer: 'Modal Footer',
  onClose: jest.fn(),
};

const OpenedModal = (
  <Modal
    title={props.title}
    isVisible={true}
    footer={[<button key="button">{props.footer}</button>]}
    onClose={props.onClose}
  >
    <p>{props.content}</p>
  </Modal>
);

type TestProps = Omit<Props, 'onClose' | 'isVisible'> & {
  onClose?: () => void;
  isVisible?: boolean;
};

function setupModal(isVisible: boolean, customProps: TestProps = {}) {
  const onClose = jest.fn();
  const ModalComponent = (
    <Modal
      title={props.title}
      isVisible={isVisible}
      footer={[<button key="button">{props.footer}</button>]}
      onClose={onClose}
      {...customProps}
    >
      <p>{props.content}</p>
    </Modal>
  );
  const { getByTestId, getByRole } = render(ModalComponent);

  return {
    modalContainer: getByTestId('modal-container') as Element,
    modalDialog: getByRole('dialog') as Element,
    closeButton: getByTestId('close-button') as Element,
    onClose,
  };
}

it('<Modal> should render with a title, content, footer and an onClick handler', () => {
  const ModalSnapshot = renderer.create(OpenedModal).toJSON();
  expect(ModalSnapshot).toMatchSnapshot();
});

describe('when modal is opened', () => {
  it('should be visible', () => {
    const { modalContainer } = setupModal(true);
    expect(modalContainer).toHaveStyle('visibility: visible');
  });

  it('should not be closed if the modal is clicked', () => {
    const { modalDialog, onClose } = setupModal(true);
    fireEvent.click(modalDialog);
    expect(onClose).toHaveBeenCalledTimes(0);
  });

  it('should display the title', () => {
    const { getByText } = render(OpenedModal);
    const modalHeader = getByText(props.title);
    expect(modalHeader).toHaveTextContent(props.title);
  });

  it('should display the content', () => {
    const { getByText } = render(OpenedModal);
    const modalContent = getByText(props.content);
    expect(modalContent).toHaveTextContent(props.content);
  });

  it('should display the footer', () => {
    const { getByText } = render(OpenedModal);
    const modalFooter = getByText(props.footer);
    expect(modalFooter).toHaveTextContent(props.footer);
  });
});

describe('when modal is closed', () => {
  it('should not be visible', () => {
    const { modalContainer } = setupModal(false);
    expect(modalContainer).toHaveStyle('visibility: hidden');
  });

  it('children should not be mounted', () => {
    const { queryByTestId } = render(
      <Modal isVisible={false} onClose={props.onClose}>
        <p data-testid="modal-children">{props.content}</p>
      </Modal>
    );
    const children = queryByTestId('modal-children');
    expect(children).not.toBeInTheDocument();
  });
});

const escapeEvent = {
  key: 'Escape',
  keyCode: 27,
};

describe('onClose should be called once when:', () => {
  it('close icon is clicked', () => {
    const { closeButton, onClose } = setupModal(true);
    fireEvent.click(closeButton);
    expect(onClose).toHaveBeenCalledTimes(1);
  });

  it('area outside modal is clicked', () => {
    const { modalContainer, onClose } = setupModal(true);
    // mousedown needs to be fired because
    // handleClick checks if the mousedown target element
    // is the modal container before calling onClose
    fireEvent.mouseDown(modalContainer);
    fireEvent.click(modalContainer);
    expect(onClose).toHaveBeenCalledTimes(1);
  });

  it('escape key is pressed', () => {
    const { modalContainer, onClose } = setupModal(true);
    fireEvent.keyDown(modalContainer, escapeEvent);
    expect(onClose).toHaveBeenCalledTimes(1);
  });
});

describe('onClose should not have been called when:', () => {
  it('escape key is pressed on closed modal', () => {
    const { modalContainer, onClose } = setupModal(false);
    fireEvent.keyDown(modalContainer, escapeEvent);
    expect(onClose).toHaveBeenCalledTimes(0);
  });

  it('onClose is not a function', () => {
    const { closeButton, onClose } = setupModal(false, {
      onClose: null,
    });
    fireEvent.click(closeButton);
    expect(onClose).toHaveBeenCalledTimes(0);
  });

  it('modal-dialog is clicked', () => {
    const { onClose } = setupModal(false);
    fireEvent.click(document.getElementsByClassName('modal-dialog')[0]);
    expect(onClose).toHaveBeenCalledTimes(0);
  });
});

describe('displays the correct size', () => {
  const sizes = Object.keys(SIZES) as sizeType[];
  sizes.forEach(size => {
    const width = SIZES[size];

    it(`${size}: ${width}`, () => {
      const { getByTestId } = render(
        <Modal size={size} isVisible={true} onClose={props.onClose}>
          <p>{props.content}</p>
        </Modal>
      );
      const modalDialog = getByTestId('dialog');
      expect(modalDialog).toHaveStyle(`width: ${width}px;`);
    });
  });
});

it('should not show the header when hideHeader is true', () => {
  const { getByTestId } = render(
    <Modal
      hideHeader
      title={props.title}
      isVisible={true}
      onClose={props.onClose}
    >
      <p>{props.content}</p>
    </Modal>
  );
  const modalContainer = getByTestId('modal-container');
  expect(modalContainer).not.toHaveTextContent(props.title);
});

it('should center the Modal vertically when centering is true', () => {
  const { getByTestId } = render(
    <Modal centering isVisible={true} onClose={props.onClose}>
      <p>{props.content}</p>
    </Modal>
  );
  const modalContainer = getByTestId('modal-container');
  expect(modalContainer).toHaveStyle('align-items: center;');
});

describe('<Modal /> not receiving prop onClose', () => {
  test('click on close button should not close the modal', () => {
    const { getByTestId } = render(
      <Modal isVisible={true}>
        <p>content</p>
      </Modal>
    );
    const modalContainer = getByTestId('modal-container');
    const closeButton = getByTestId('close-button');
    fireEvent.click(closeButton);
    expect(modalContainer).toHaveStyle('visibility: visible');
  });
  test('click outside should not close the modal', () => {
    const { queryByTestId, queryByText } = render(
      <React.Fragment>
        <Modal isVisible={true}>
          <p>content</p>
        </Modal>
        <div>outside</div>
      </React.Fragment>
    );
    const modalContainer = queryByTestId('modal-container');
    const outside = queryByText('outside');
    fireEvent.click(outside);
    expect(modalContainer).toHaveStyle('visibility: visible');
  });
});
