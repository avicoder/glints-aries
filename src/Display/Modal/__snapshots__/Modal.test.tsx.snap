// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Modal> should match snapshot 1`] = `
<div
  className="aries-modal ModalStyle__ModalContainer-sc-101ur27-0 gsOTFp"
  data-testid="modal-container"
  onClick={[Function]}
  onMouseDown={[Function]}
  onTransitionEnd={[Function]}
>
  <div
    className="modal-dialog ModalStyle__ModalDialog-sc-101ur27-1 ShNNk"
  >
    <div
      aria-modal="true"
      className="modal-content ModalStyle__ModalContentArea-sc-101ur27-2 dySsLM"
      onClick={[Function]}
      role="dialog"
      tabIndex={0}
    >
      <header
        className="modal-header ModalStyle__ModalHeader-sc-101ur27-3 eEgWgO"
      >
        <h3>
          Modal Title
        </h3>
        <button
          data-testid="close-button"
          onClick={[MockFunction]}
          type="button"
        >
          <svg
            data-testid="icon-svg"
            fill="grey"
            height="1em"
            viewBox="0 0 100 100"
            width="1em"
          >
            <path
              d="M100 10L90 0 50 40 10 0 0 10l40 40L0 90l10 10 40-40 40 40 10-10-40-40z"
            />
          </svg>
        </button>
      </header>
      <section
        className="modal-body ModalStyle__ModalBody-sc-101ur27-4 iWAGvJ"
      >
        <p>
          Modal Content
        </p>
      </section>
      <footer
        className="modal-footer ModalStyle__ModalFooter-sc-101ur27-5 dNhZMD"
      >
        <button>
          Modal Footer
        </button>
      </footer>
    </div>
  </div>
</div>
`;
